File: apps/desktop/renderer/src/workspace.ts
/**
 * Pure reducers for the sidebar's local organisation: spaces, tab order,
 * pinned sites. The runtime owns pages; the shell owns how they are arranged.
 * Everything here is deterministic and unit-tested.
 */
import type { PageTarget } from "@vector/contracts";

export type SpaceColor = "blue" | "violet" | "pink" | "orange" | "green" | "teal" | "slate";
export const SPACE_COLORS: SpaceColor[] = ["blue", "violet", "pink", "orange", "green", "teal", "slate"];

export interface Space {
  id: string;
  name: string;
  color: SpaceColor;
}

export interface Pin {
  url: string;
  title: string;
}

export const MAX_PINS = 5;

export interface Folder {
  id: string;
  name: string;
  spaceId: string;
  collapsed: boolean;
}

export interface Layout {
  spaces: Space[];
  activeSpaceId: string;
  /** pageId → spaceId; pages not listed belong to the active space when first seen */
  tabSpace: Record<string, string>;
  /** explicit order of pageIds within the sidebar (across spaces) */
  order: string[];
  /** pinned sites per space */
  pins: Record<string, Pin[]>;
  /** named tab folders within a space (Arc-style) */
  folders: Folder[];
  /** pageId → folderId; missing means unfiled (shown below New Tab) */
  tabFolder: Record<string, string>;
}

export const DEFAULT_SPACE: Space = { id: "space-personal", name: "Personal", color: "blue" };

export function emptyLayout(): Layout {
  return { spaces: [DEFAULT_SPACE], activeSpaceId: DEFAULT_SPACE.id, tabSpace: {}, order: [], pins: { [DEFAULT_SPACE.id]: [] }, folders: [], tabFolder: {} };
}

let seq = 0;
export const newSpaceId = () => `space-${Date.now().toString(36)}-${(seq++).toString(36)}`;
export const newFolderId = () => `folder-${Date.now().toString(36)}-${(seq++).toString(36)}`;

/** Tabs the sidebar shows: human-facing, not worker pages. */
export function isTab(p: Pick<PageTarget, "viewStatus" | "ownedByRuntime">): boolean {
  return p.viewStatus !== "background" && p.viewStatus !== "detached" && !p.ownedByRuntime;
}

/** Bring `order` in sync with the live page list: keep known order, append new pages, drop closed. */
export function syncOrder(order: string[], pages: Pick<PageTarget, "pageId">[]): string[] {
  const live = new Set(pages.map((p) => p.pageId));
  const kept = order.filter((id) => live.has(id));
  const known = new Set(kept);
  for (const p of pages) if (!known.has(p.pageId)) kept.push(p.pageId);
  return kept;
}

/** Ensure every live page has a space; new pages land in the active space. */
export function syncSpaces(layout: Layout, pages: Pick<PageTarget, "pageId">[]): Layout {
  const tabSpace: Record<string, string> = {};
  const valid = new Set(layout.spaces.map((s) => s.id));
  let changed = false;
  for (const p of pages) {
    const cur = layout.tabSpace[p.pageId];
    const next = cur && valid.has(cur) ? cur : layout.activeSpaceId;
    if (next !== cur) changed = true;
    tabSpace[p.pageId] = next;
  }
  if (!changed && Object.keys(layout.tabSpace).length === Object.keys(tabSpace).length) return layout;
  return { ...layout, tabSpace };
}

/** Ordered tabs for one space. */
export function tabsInSpace<T extends Pick<PageTarget, "pageId" | "viewStatus" | "ownedByRuntime">>(
  layout: Layout,
  pages: T[],
  spaceId: string,
): T[] {
  const byId = new Map(pages.map((p) => [p.pageId, p]));
  const out: T[] = [];
  for (const id of layout.order) {
    const p = byId.get(id);
    if (p && isTab(p) && (layout.tabSpace[id] ?? layout.activeSpaceId) === spaceId) out.push(p);
  }
  // pages that have not been ordered yet (racing an event) trail the list
  for (const p of pages) {
    if (!layout.order.includes(p.pageId) && isTab(p) && (layout.tabSpace[p.pageId] ?? layout.activeSpaceId) === spaceId) out.push(p);
  }
  return out;
}

/** Move a tab to sit before `beforeId` (or at the end when null), keeping other spaces untouched. */
export function moveTab(layout: Layout, pageId: string, beforeId: string | null): Layout {
  if (pageId === beforeId) return layout;
  const without = layout.order.filter((id) => id !== pageId);
  if (beforeId === null) return { ...layout, order: [...without, pageId] };
  const idx = without.indexOf(beforeId);
  if (idx === -1) return { ...layout, order: [...without, pageId] };
  without.splice(idx, 0, pageId);
  return { ...layout, order: without };
}

/** Reorder by index within a space — what a drag gesture produces. */
export function reorderInSpace(layout: Layout, spaceId: string, from: number, to: number, pages: Pick<PageTarget, "pageId" | "viewStatus" | "ownedByRuntime">[]): Layout {
  const tabs = tabsInSpace(layout, pages, spaceId).map((p) => p.pageId);
  if (from < 0 || from >= tabs.length || to < 0 || to > tabs.length || from === to) return layout;
  const moving = tabs[from]!;
  const rest = tabs.filter((_, i) => i !== from);
  const insertAt = to > from ? to - 1 : to;
  rest.splice(insertAt, 0, moving);
  // rebuild the global order: replace this space's slots with the new sequence
  const slots = new Set(tabs);
  let k = 0;
  const order = layout.order.map((id) => (slots.has(id) ? rest[k++]! : id));
  for (; k < rest.length; k++) order.push(rest[k]!);
  return { ...layout, order };
}

export function assignSpace(layout: Layout, pageId: string, spaceId: string): Layout {
  if (!layout.spaces.some((s) => s.id === spaceId)) return layout;
  const tabSpace = { ...layout.tabSpace, [pageId]: spaceId };
  const fid = layout.tabFolder[pageId];
  const folder = fid ? layout.folders.find((f) => f.id === fid) : undefined;
  if (folder && folder.spaceId !== spaceId) {
    const tabFolder = { ...layout.tabFolder };
    delete tabFolder[pageId];
    return { ...layout, tabSpace, tabFolder };
  }
  return { ...layout, tabSpace };
}

export function createSpace(layout: Layout, name: string, color?: SpaceColor): Layout {
  const used = new Set(layout.spaces.map((s) => s.color));
  const pick = color ?? SPACE_COLORS.find((c) => !used.has(c)) ?? SPACE_COLORS[layout.spaces.length % SPACE_COLORS.length]!;
  const space: Space = { id: newSpaceId(), name: name.trim() || `Space ${layout.spaces.length + 1}`, color: pick };
  return { ...layout, spaces: [...layout.spaces, space], activeSpaceId: space.id, pins: { ...layout.pins, [space.id]: [] } };
}

export function renameSpace(layout: Layout, spaceId: string, name: string, color?: SpaceColor): Layout {
  return {
    ...layout,
    spaces: layout.spaces.map((s) => (s.id === spaceId ? { ...s, name: name.trim() || s.name, color: color ?? s.color } : s)),
  };
}

/** Removing a space folds its tabs and pins into the neighbouring space. The last space cannot be removed. */
export function removeSpace(layout: Layout, spaceId: string): Layout {
  if (layout.spaces.length <= 1) return layout;
  const idx = layout.spaces.findIndex((s) => s.id === spaceId);
  if (idx === -1) return layout;
  const fallback = layout.spaces[idx === 0 ? 1 : idx - 1]!;
  const tabSpace = Object.fromEntries(Object.entries(layout.tabSpace).map(([k, v]) => [k, v === spaceId ? fallback.id : v]));
  const pins = { ...layout.pins };
  const moved = pins[spaceId] ?? [];
  delete pins[spaceId];
  pins[fallback.id] = dedupePins([...(pins[fallback.id] ?? []), ...moved]);
  const folders = layout.folders.map((f) => (f.spaceId === spaceId ? { ...f, spaceId: fallback.id } : f));
  return {
    ...layout,
    spaces: layout.spaces.filter((s) => s.id !== spaceId),
    activeSpaceId: layout.activeSpaceId === spaceId ? fallback.id : layout.activeSpaceId,
    tabSpace,
    pins,
    folders,
  };
}

export function setActiveSpace(layout: Layout, spaceId: string): Layout {
  return layout.spaces.some((s) => s.id === spaceId) ? { ...layout, activeSpaceId: spaceId } : layout;
}

export function togglePin(layout: Layout, spaceId: string, pin: Pin): Layout {
  const cur = layout.pins[spaceId] ?? [];
  const has = cur.some((p) => p.url === pin.url);
  const next = has ? cur.filter((p) => p.url !== pin.url) : [...cur, pin].slice(0, MAX_PINS);
  return { ...layout, pins: { ...layout.pins, [spaceId]: next } };
}

export function isPinned(layout: Layout, spaceId: string, url: string): boolean {
  return (layout.pins[spaceId] ?? []).some((p) => p.url === url);
}

export function foldersInSpace(layout: Layout, spaceId: string): Folder[] {
  return layout.folders.filter((f) => f.spaceId === spaceId);
}

export function tabsInFolder<T extends Pick<PageTarget, "pageId" | "viewStatus" | "ownedByRuntime">>(
  layout: Layout,
  pages: T[],
  folderId: string,
): T[] {
  return tabsInSpace(layout, pages, layout.folders.find((f) => f.id === folderId)?.spaceId ?? "").filter((p) => layout.tabFolder[p.pageId] === folderId);
}

export function unfiledTabs<T extends Pick<PageTarget, "pageId" | "viewStatus" | "ownedByRuntime">>(
  layout: Layout,
  pages: T[],
  spaceId: string,
): T[] {
  const ids = new Set(foldersInSpace(layout, spaceId).map((f) => f.id));
  return tabsInSpace(layout, pages, spaceId).filter((p) => {
    const fid = layout.tabFolder[p.pageId];
    return !fid || !ids.has(fid);
  });
}

export function createFolder(layout: Layout, spaceId: string, name: string): Layout {
  if (!layout.spaces.some((s) => s.id === spaceId)) return layout;
  const folder: Folder = { id: newFolderId(), name: name.trim() || "Untitled", spaceId, collapsed: false };
  return { ...layout, folders: [...layout.folders, folder] };
}

export function renameFolder(layout: Layout, folderId: string, name: string): Layout {
  const n = name.trim();
  if (!n) return layout;
  return { ...layout, folders: layout.folders.map((f) => (f.id === folderId ? { ...f, name: n } : f)) };
}

export function toggleFolder(layout: Layout, folderId: string): Layout {
  return { ...layout, folders: layout.folders.map((f) => (f.id === folderId ? { ...f, collapsed: !f.collapsed } : f)) };
}

export function removeFolder(layout: Layout, folderId: string): Layout {
  const tabFolder = { ...layout.tabFolder };
  for (const [pageId, fid] of Object.entries(tabFolder)) if (fid === folderId) delete tabFolder[pageId];
  return { ...layout, folders: layout.folders.filter((f) => f.id !== folderId), tabFolder };
}

export function assignFolder(layout: Layout, pageId: string, folderId: string | null): Layout {
  const tabFolder = { ...layout.tabFolder };
  if (!folderId) {
    delete tabFolder[pageId];
    return { ...layout, tabFolder };
  }
  if (!layout.folders.some((f) => f.id === folderId)) return layout;
  tabFolder[pageId] = folderId;
  return { ...layout, tabFolder };
}

/** Reorder tabs that currently display as one group (a folder, or the unfiled list). */
export function reorderGroup(
  layout: Layout,
  spaceId: string,
  folderId: string | null,
  from: number,
  to: number,
  pages: Pick<PageTarget, "pageId" | "viewStatus" | "ownedByRuntime">[],
): Layout {
  const group = folderId ? tabsInFolder(layout, pages, folderId) : unfiledTabs(layout, pages, spaceId);
  const ids = group.map((p) => p.pageId);
  if (from < 0 || from >= ids.length || to < 0 || to > ids.length || from === to) return layout;
  const moving = ids[from]!;
  const rest = ids.filter((_, i) => i !== from);
  const insertAt = to > from ? to - 1 : to;
  rest.splice(insertAt, 0, moving);
  const slots = new Set(ids);
  let k = 0;
  const order = layout.order.map((id) => (slots.has(id) ? rest[k++]! : id));
  for (; k < rest.length; k++) order.push(rest[k]!);
  return { ...layout, order };
}

function dedupePins(pins: Pin[]): Pin[] {
  const seen = new Set<string>();
  return pins.filter((p) => (seen.has(p.url) ? false : (seen.add(p.url), true)));
}

/** Round-trip through storage — tolerant of missing/corrupt data. */
export function parseLayout(raw: string | null): Layout {
  if (!raw) return emptyLayout();
  try {
    const v = JSON.parse(raw) as Partial<Layout>;
    if (!Array.isArray(v.spaces) || v.spaces.length === 0) return emptyLayout();
    const spaces = v.spaces.filter((s): s is Space => !!s && typeof s.id === "string" && typeof s.name === "string");
    if (!spaces.length) return emptyLayout();
    const active = spaces.some((s) => s.id === v.activeSpaceId) ? v.activeSpaceId! : spaces[0]!.id;
    return {
      spaces,
      activeSpaceId: active,
      tabSpace: v.tabSpace && typeof v.tabSpace === "object" ? v.tabSpace : {},
      order: Array.isArray(v.order) ? v.order.filter((x): x is string => typeof x === "string") : [],
      pins: v.pins && typeof v.pins === "object"
        ? Object.fromEntries(
            Object.entries(v.pins).map(([id, list]) => [id, Array.isArray(list) ? (list as Pin[]).slice(0, MAX_PINS) : []]),
          )
        : {},
      folders: Array.isArray(v.folders)
        ? v.folders.flatMap((f) => {
            if (!f || typeof f.id !== "string" || typeof f.name !== "string" || typeof f.spaceId !== "string") return [];
            return [{ id: f.id, name: f.name, spaceId: f.spaceId, collapsed: !!f.collapsed }];
          })
        : [],
      tabFolder: v.tabFolder && typeof v.tabFolder === "object"
        ? Object.fromEntries(Object.entries(v.tabFolder).filter((entry): entry is [string, string] => typeof entry[1] === "string"))
        : {},
    };
  } catch {
    return emptyLayout();
  }
}

/** Host without www — used for tiles, pins, and the start page. */
export function hostOf(url: string): string {
  try {
    return new URL(url).hostname.replace(/^www\./, "");
  } catch {
    return url;
  }
}

File: engine/crates/ve-chrome/src/workspace.rs
//! Port of `apps/desktop/renderer/src/workspace.ts`.

use serde::{Deserialize, Serialize};

/// Space identity colours (grayscale steps in the token sheet).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SpaceColor {
    /// Blue slot.
    Blue,
    /// Violet slot.
    Violet,
    /// Pink slot.
    Pink,
    /// Orange slot.
    Orange,
    /// Green slot.
    Green,
    /// Teal slot.
    Teal,
    /// Slate slot.
    Slate,
}

/// One Arc-style space.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Space {
    /// Stable id.
    pub id: String,
    /// Display name.
    pub name: String,
    /// Identity colour.
    pub color: SpaceColor,
}

/// Pinned site (cap 5).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Pin {
    /// URL.
    pub url: String,
    /// Title.
    pub title: String,
}

/// Maximum pins per space.
pub const MAX_PINS: usize = 5;

/// Named tab folder.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Folder {
    /// Stable id.
    pub id: String,
    /// Display name.
    pub name: String,
    /// Owning space.
    pub space_id: String,
    /// Collapsed in the sidebar.
    pub collapsed: bool,
}

/// Sidebar organisation. The runtime owns pages; chrome owns arrangement.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Layout {
    /// Spaces.
    pub spaces: Vec<Space>,
    /// Active space id.
    pub active_space_id: String,
    /// page id → space id.
    pub tab_space: Vec<(String, String)>,
    /// Explicit tab order.
    pub order: Vec<String>,
    /// Pins per space.
    pub pins: Vec<(String, Vec<Pin>)>,
    /// Folders.
    pub folders: Vec<Folder>,
    /// page id → folder id.
    pub tab_folder: Vec<(String, String)>,
}

/// Default Personal space.
#[must_use]
pub fn default_space() -> Space {
    Space {
        id: "space-personal".into(),
        name: "Personal".into(),
        color: SpaceColor::Blue,
    }
}

/// Empty layout with one space.
#[must_use]
pub fn empty_layout() -> Layout {
    let space = default_space();
    Layout {
        active_space_id: space.id.clone(),
        pins: vec![(space.id.clone(), Vec::new())],
        spaces: vec![space],
        tab_space: Vec::new(),
        order: Vec::new(),
        folders: Vec::new(),
        tab_folder: Vec::new(),
    }
}

/// Host without `www.` — tiles, pins, start page.
#[must_use]
pub fn host_of(url: &str) -> String {
    url::Url::parse(url)
        .ok()
        .and_then(|u| u.host_str().map(str::to_string))
        .map(|h| h.trim_start_matches("www.").to_string())
        .unwrap_or_else(|| url.to_string())
}

/// Keep `order` in sync with live page ids.
#[must_use]
pub fn sync_order(order: &[String], pages: &[String]) -> Vec<String> {
    let live: std::collections::HashSet<&str> = pages.iter().map(String::as_str).collect();
    let mut kept: Vec<String> = order
        .iter()
        .filter(|id| live.contains(id.as_str()))
        .cloned()
        .collect();
    let known: std::collections::HashSet<String> = kept.iter().cloned().collect();
    for p in pages {
        if !known.contains(p) {
            kept.push(p.clone());
        }
    }
    kept
}

/// Assign unknown pages to the active space.
#[must_use]
pub fn sync_spaces(layout: &Layout, pages: &[String]) -> Layout {
    let valid: std::collections::HashSet<&str> =
        layout.spaces.iter().map(|s| s.id.as_str()).collect();
    let mut tab_space = Vec::new();
    for p in pages {
        let cur = layout
            .tab_space
            .iter()
            .find(|(id, _)| id == p)
            .map(|(_, s)| s.as_str());
        let next = cur
            .filter(|s| valid.contains(*s))
            .unwrap_or(layout.active_space_id.as_str());
        tab_space.push((p.clone(), next.to_string()));
    }
    Layout {
        tab_space,
        ..layout.clone()
    }
}

/// Ordered tabs for one space.
#[must_use]
pub fn tabs_in_space(layout: &Layout, pages: &[String], space_id: &str) -> Vec<String> {
    let mut out = Vec::new();
    for id in &layout.order {
        if pages.iter().any(|p| p == id) {
            let space = layout
                .tab_space
                .iter()
                .find(|(p, _)| p == id)
                .map(|(_, s)| s.as_str())
                .unwrap_or(layout.active_space_id.as_str());
            if space == space_id {
                out.push(id.clone());
            }
        }
    }
    for p in pages {
        if !layout.order.contains(p) {
            let space = layout
                .tab_space
                .iter()
                .find(|(id, _)| id == p)
                .map(|(_, s)| s.as_str())
                .unwrap_or(layout.active_space_id.as_str());
            if space == space_id {
                out.push(p.clone());
            }
        }
    }
    out
}

/// Pin or unpin a URL in a space (cap [`MAX_PINS`]).
#[must_use]
pub fn toggle_pin(layout: &Layout, space_id: &str, pin: Pin) -> Layout {
    let mut pins = layout.pins.clone();
    let entry = pins.iter_mut().find(|(id, _)| id == space_id);
    if let Some((_, list)) = entry {
        if let Some(i) = list.iter().position(|p| p.url == pin.url) {
            list.remove(i);
        } else {
            list.push(pin);
            list.truncate(MAX_PINS);
        }
    } else {
        pins.push((space_id.to_string(), vec![pin]));
    }
    Layout {
        pins,
        ..layout.clone()
    }
}

/// Switch the active space.
#[must_use]
pub fn set_active_space(layout: &Layout, space_id: &str) -> Layout {
    if layout.spaces.iter().any(|s| s.id == space_id) {
        Layout {
            active_space_id: space_id.to_string(),
            ..layout.clone()
        }
    } else {
        layout.clone()
    }
}

/// Parse a stored layout; corrupt data becomes [`empty_layout`].
#[must_use]
pub fn parse_layout(raw: Option<&str>) -> Layout {
    let Some(raw) = raw else {
        return empty_layout();
    };
    serde_json::from_str(raw).unwrap_or_else(|_| empty_layout())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sync_order_keeps_and_appends() {
        let order = sync_order(&["b".into(), "a".into()], &["a".into(), "c".into()]);
        assert_eq!(order, vec!["a".to_string(), "c".to_string()]);
    }

    #[test]
    fn host_strips_www() {
        assert_eq!(host_of("https://www.example.test/x"), "example.test");
    }

    #[test]
    fn pin_cap() {
        let mut layout = empty_layout();
        for i in 0..6 {
            layout = toggle_pin(
                &layout,
                &layout.active_space_id,
                Pin {
                    url: format!("https://p{i}.test/"),
                    title: format!("{i}"),
                },
            );
        }
        let n = layout.pins[0].1.len();
        assert_eq!(n, MAX_PINS);
    }
}
